src/input.rs

```diff
@@ -64,25 +73,87 @@ impl GraphConverter {
         let padding_id = self.add_or_get_val(padding);
         let activation_id = self.add_or_get_val(activation);
 
-        let conv_node = Mdl::Conv2d([
+        let new_node = Mdl::Conv2d([
             stride_h_id,
             stride_w_id,
             padding_id,
             activation_id,
             inpt,
             wght,
         ]);
-        self.rec_expr.add(conv_node)
+        self.rec_expr.add(new_node)
     }
 
     pub fn relu(&mut self, inpt: Id) -> Id {
-        let relu_node = Mdl::Relu(inpt);
-        self.rec_expr.add(relu_node)
+        let new_node = Mdl::Relu(inpt);
+        self.rec_expr.add(new_node)
+    }
+
+    pub fn tanh(&mut self, inpt: Id) -> Id {
+        let new_node = Mdl::Tanh(inpt);
+        self.rec_expr.add(new_node)
+    }
+
+    pub fn sigmoid(&mut self, inpt: Id) -> Id {
+        let new_node = Mdl::Sigmoid(inpt);
+        self.rec_expr.add(new_node)
     }
 
     pub fn add(&mut self, inpt_1: Id, inpt_2: Id) -> Id {
-        let add_node = Mdl::Ewadd([inpt_1, inpt_2]);
-        self.rec_expr.add(add_node)
+        let new_node = Mdl::Ewadd([inpt_1, inpt_2]);
+        self.rec_expr.add(new_node)
+    }
+
+    pub fn matmul(&mut self, inpt_1: Id, inpt_2: Id) -> Id {
+        let activation = ACTNONE;
+        let act_id = self.add_or_get_val(activation);
+
+        let new_node = Mdl::Matmul([act_id, inpt_1, inpt_2]);
+        self.rec_expr.add(new_node)
+    }
+
+    pub fn mul(&mut self, inpt_1: Id, inpt_2: Id) -> Id {
+        let new_node = Mdl::Ewmul([inpt_1, inpt_2]);
+        self.rec_expr.add(new_node)
+    }
+
+    pub fn concat(&mut self, axis: i32, ndim: i32, inpt_1: Id, inpt_2: Id) -> Id {
+        // Only support concat of 2 inputs for now
+        // To support more, pass in a slice and create more concat nodes here
```

Comment:
Why does one need to pass ndim here? (at least ndim to me sounds like rank of inpt_1 and inpt_2)

Reply:
inpt_1 and inpt_2 are both of type Id, which is just the index into RecExpr. It doesn't contain any tensor information. So I need to pass in the information explicitly. ndim is needed for concat nodes in egg since the rewrite rules can require to match only concat with a specific ndim.